static/js/compose_ui.js

```diff
@@ -221,17 +221,77 @@ export function handle_keyup(event, textarea) {
 }
 
 export function format_text(textarea, type) {
+    const italic_syntax = "*";
+    const bold_syntax = "**";
+    const bold_and_italic_syntax = "***";
+    let is_selected_text_italic = false;
     const field = textarea.get(0);
     const range = textarea.range();
+    let text = textarea.val();
 
     switch (type) {
         case "bold":
+            // If the text is already bold, we remove the bold_syntax from text.
+            // Check if characters around selected text have enough length to have bold_syntax.
+            if (
+                range.start >= 2 &&
+                text.length - range.end >= 2 && // If the characters before and after selected text have bold_syntax.
+                text.slice(range.start - 2, range.start) === bold_syntax &&
+                text.slice(range.end, range.end + 2) === bold_syntax
+            ) {
```

Comment:
One thing that I notice is that this works if you have `**[selected text]**`, but not `[**selected text**]`, which seems like something a user would do if they were trying to remove the bold a while after adding it in the first place.  GitHub seems to handle that case, so seems like we should too.

The other thing I notice is that this logic could probably be deduplicated with the italics logic; the `2` here is just `syntax.length` and the main thing that differs.

This seems worth doing since we may also decide to add `~~strikethrough~~` and code block wrapping to what's supported by this feature.

Reply:
Fixed it and reduced some duplecacy.

Reply:
Here's a readability refactoring I had in mind, replacing the `1/2` with `.length`, which might also help with reusability:


```diff --git a/static/js/compose_ui.js b/static/js/compose_ui.js
index b1319b2818..460ab4ec98 100644
--- a/static/js/compose_ui.js
+++ b/static/js/compose_ui.js
@@ -231,21 +231,24 @@ export function format_text(textarea, type) {
 
     switch (type) {
         case "bold":
-            // If the text is already bold, we remove the bold_syntax from text.
-            // Check if characters around selected text have enough length to have bold_syntax.
+            // If the selection is already surrounded by bold syntax,
+            // remove it rather than adding another copy.
             if (
-                range.start >= 2 &&
-                text.length - range.end >= 2 && // If the characters before and after selected text have bold_syntax.
-                text.slice(range.start - 2, range.start) === bold_syntax &&
-                text.slice(range.end, range.end + 2) === bold_syntax
+                // First check if there are enough characters before/after selection.
+                range.start >= bold_syntax.length &&
+                text.length - range.end >= bold_syntax.length &&
+                // And then if the characters are bold_syntax.
+                text.slice(range.start - bold_syntax.length, range.start) === bold_syntax &&
+                text.slice(range.end, range.end + bold_syntax.length) === bold_syntax
             ) {
                 // Remove the bold_syntax from text.
                 text =
-                    text.slice(0, range.start - 2) +
+                    text.slice(0, range.start - bold_syntax.length) +
                     text.slice(range.start, range.end) +
-                    text.slice(range.end + 2);
+                    text.slice(range.end + bold_syntax.length);
                 textarea.val(text);
-                field.setSelectionRange(range.start - 2, range.end - 2);
+                field.setSelectionRange(range.start - bold_syntax.length,
+                                        range.end - bold_syntax.length);
                 break;
             }
 

```

Reply:
Done.